src/Core/Test/PriorityProducerConsumerTest.cs

```diff
@@ -178,8 +178,10 @@ public class PriorityProducerConsumerTest {
                 for (var i = 0; i < 3; i++) {
                     var task = ppc.ConsumeAsync();
                     tcsConsumer.TrySetResult(true);
-                    values[i] = await task;
+
+                    // make sure wait for producer first to remove race
                     await tcsProducer.Task;
+                    values[i] = await task;
```

Comment:
this is where behavior got changed. previously if ConsumerAsync is called when the queue is empty, caller will hold onto TaskCompletionSource<T>.Task of pendingTcs directly

```csharp
                pendingTcs = new TaskCompletionSource<T>(TaskCreationOptions.RunContinuationsAsynchronously);	
                pending = new Pending(pendingTcs, _syncObj);	
                _pendingTasks.Enqueue(pending);	
            }	

            return pendingTcs.Task;
```

and when finally something enqueued through Produce

```csharp
            pendingTcs?.TrySetResult(value);
```

code set the value explicitly to the source, making the ConsumerAsync caller to get that value right away.

that is why existing test get the value in this order

```csharp
			Assert.AreEqual(5, values[0]);	
            Assert.AreEqual(7, values[0]);
            Assert.AreEqual(6, values[2]);
```

with the change, now ConsumeAsync always gets value in priority order among values at the moment in the queue, so now it gets in order of 7,5,6.

also, I moved await in the test to make sure test retrieve values once all Produce is called otherwise, we could have nondeterministic result since values are based on priority at the moment in the queue.

```csharp
	await tcsProducer.Task;
```

Reply:
Ok, so here is the behavior of the original test:
- Step 1: one consumer task is registered, no tasks in the queue
- Step 2: item with priority 1 pushed to consumer. Since there is no item with priority 0, it is set to consumer. But consumer awaits for `tcsConsumer.Task`, so no new waiter are registered
- Step 3: items with priority 1 and 0 are registered, and `tcsProducer` is set
- Step 4: `consumerTask` drains all the items with respect of the order

This is an expected behavior, cause in ppc use cases, tasks with lower priority happen more often. For instance, in Language Server, initialization tasks have priority 0, configuration change tasks have priority 1, and document change tasks have priority 2.

Reply:
@AlexanderSher 
so, consumer task that created when queue is empty, must get the very first item added to the queue when it was empty? is that an explicit contract or implementation contract code assumes in other parts of code base?

basically, what I am asking is if I change that behavior, will it break something?

the change I made basically make "consumer task" that created when queue is empty, will get an item in the queue once someone put items in the queue based on priority

Reply:
> basically, what I am asking is if I change that behavior, will it break something?

**If** I correctly understand your changes, it will be a breaking change. PPC is used in LS to prioritize requests when they come faster than we process them, but when the queue is empty, consumer awaits for any incoming request, e.g. it doesn't wait for request with priority 0 to start processing requests with priority 1 and 2.

Reply:
@AlexanderSher the change still doesn't wait for priority 0? it just wont special case the empty case anymore. the queue will pick up next time in the queue and process it as all other cases.

in reality, how the system knows this specific situation?